### PR TITLE
HW#2: add documentation for Stack/StackStatistics classes.

### PR DESCRIPTION
**Pull request by pingvin1977**

### CODE REVIEW

File: iOS-School-02_stack.playground/Contents.swift [hunk 1]
```diff
@@ -1,33 +1,78 @@
+//MARK: - Stack
+/**
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:2]: Leave a blank line after `// MARK:`

File: iOS-School-02_stack.playground/Contents.swift [hunk 2]
```diff
@@ -1,33 +1,78 @@
+//MARK: - Stack
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:1]:
```suggestion
//  MARK: - Stack
```

Check the rest please

File: iOS-School-02_stack.playground/Contents.swift [hunk 3]
```diff
@@ -1,33 +1,78 @@
...
+/**
+ Provides a standard interface for on a collection,
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:3]: > for on a collection

What does that mean?

File: iOS-School-02_stack.playground/Contents.swift [hunk 4]
```diff
@@ -1,33 +1,78 @@
...
+ wich is organized according to the principle "last in, first out", called a
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:4]:
```suggestion
 which is organized according to the principle "last in, first out", called a
```

File: iOS-School-02_stack.playground/Contents.swift [hunk 5]
```diff
@@ -1,33 +1,78 @@
...
+ [stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)).
+ The stack can hold `Int` values only.
+ You can add, remove an element from the stack and peek the element that added last.
+ */
 class Stack {
+    /// Contains element of the stack.
     private var stackElements = [Int]()
-
+    /**
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:12]: Leave a blank after the last declaration (property/function).
Please check the rest.

File: iOS-School-02_stack.playground/Contents.swift (hunk 4 above)
Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift:4]: Why you didn't use grammarly service?

File: iOS-School-02_stack.playground/Contents.swift [hunk 6]
```diff
@@ -1,33 +1,78 @@
...
+     Adds the element to the stack.
+     - Parameter element; An element to add on the stack.
+     - Complexity: 0(1)
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:15]:
```suggestion
     - Complexity: O(1)
```

Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift:15]: Fix the rest.

File: iOS-School-02_stack.playground/Contents.swift [hunk 7]
```diff
@@ -1,33 +1,78 @@
...
+     */
     func pop() -> Int? {
         return stackElements.popLast()
     }
-
+    /**
+     Returns the element that added last to the stack, but doesn't remove it.
+     - Returns: The element that added last to the stack. If the stack is empty, returns `nil`.
+     - Complexity: 0(1)
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:32]:
```suggestion
     - Complexity: O(1)
```

File: iOS-School-02_stack.playground/Contents.swift [hunk 8]
```diff
@@ -1,33 +1,78 @@
...
+/**
+ Provides a standard interface for operation on a collection,
+ which is organized according to the principle "last in, first out", called a
+ [stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)).
+ The stack can hold `Int` values only.
+ You can add, remove an element from the stack, peek the element that added last
+ and get the minimum value in the stack.
+ */
 class StackStatistics: Stack {
+    ///
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:48]: ?

File: iOS-School-02_stack.playground/Contents.swift [hunk 9]
```diff
@@ -1,33 +1,78 @@
...
     private var currentMinElements = [Int]()
-
+    /**
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:50]: Why you're copy-pasting this docs? Is that required?

File: iOS-School-02_stack.playground/Contents.swift [hunk 10]
```diff
@@ -1,33 +1,78 @@
...
+     - Parameter element: An element to add on the stack.
+     - Complexity: 0(1)
+     */
     override func push(_ element: Int) {
         super.push(element)
         let minimunValue = min(element, currentMinElements.last ?? element)
         currentMinElements.append(minimunValue)
     }
-
+    /**
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:60]: Why you're copy-pasting this docs? Is that required?

**Review by dersim-davaod — CHANGES_REQUESTED** [review]

File: iOS-School-02_stack.playground/Contents.swift (hunk 1 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:2]: Added lines.

File: iOS-School-02_stack.playground/Contents.swift (hunk 2 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:1]: Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 4 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:4]: Grammarly service does not check grammar in Xcode.

File: iOS-School-02_stack.playground/Contents.swift (hunk 6 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:15]: Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 7 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:32]: Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 5 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:12]: Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 3 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:3]: Sorry, i missed a word "operation". Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 9 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:50]: I didn't leave blank lines. Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 10 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:60]: I didn't leave blank lines. Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 8 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift]: Added description.

File: iOS-School-02_stack.playground/Contents.swift (hunk 4 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift]: Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 2 above)
Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift]: Sorry, I did a typo there.
I meant to have only one space after `//` characters.

Please fix it.

File: iOS-School-02_stack.playground/Contents.swift (hunk 4 above)
Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift]: Sure, it doesn't! There's no code editor that does this.
You can copy-paste some text and check the grammar/spelling. Then copy text back into Xcode.

Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift]: My point as we discussed - this is your job to make sure grammar is ok before submitting PR.

File: iOS-School-02_stack.playground/Contents.swift (hunk 5 above)
Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift:14]: No it doesn't. Line 22, for example:
```swift
    func push(_ element: Int) {
        stackElements.append(element)		    
        stackElements.append(element)	    
    }
    /**
    Removes the element that added last from the stack and return it.

```

File: iOS-School-02_stack.playground/Contents.swift (hunk 9 above)
Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift:55]: I mean why you have copied the same docs from `Stack` class? Is that required?

File: iOS-School-02_stack.playground/Contents.swift [hunk 11]
```diff
@@ -1,33 +1,85 @@
...
+ Provides a standard interface for operation on a collection,
+ which is organized according to the principle "last in, first out", called a
+ [stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)).
+ The stack can hold `Int` values only.
+ You can add, remove an element from the stack, peek the element that added last
+ and get the minimum value in the stack.
+ */
 class StackStatistics: Stack {
+    /// Sorted stack from maximum to minimum.
+    /// The last added value is the minimum value on the stack.
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:52]: > Sorted stack from maximum to minimum.

So if our stack is

`[1,5,7,3,5,9]`

, then `currentMinElements` will be

`[9,7,5,5,3,1]`

Am I right?

**Review by dersim-davaod — CHANGES_REQUESTED** [review]

File: iOS-School-02_stack.playground/Contents.swift (hunk 2 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift]: Fixed.

File: iOS-School-02_stack.playground/Contents.swift (hunk 5 above)
Comment (pingvin1977) [reply iOS-School-02_stack.playground/Contents.swift:14]: Fixed.

File: iOS-School-02_stack.playground/Contents.swift [hunk 12]
```diff
@@ -1,20 +1,57 @@
...
+/**
+ Provides a standard interface for operation on a collection,
+ which is organized according to the principle "last in, first out", called a
+ [stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)).
+ The stack can hold `Int` values only.
+ You can add, remove an element from the stack, peek the element that added last
+ and get the minimum value in the stack.
+ */
 class StackStatistics: Stack {
+    /// Contains of minimal elements.
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:54]: 1. Check grammar!

Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:54]: > Contains of minimal elements.

So, if our stack is

[1,5,7,3,5,1,9]

, then currentMinElements will be

[1,1] as this is min elements.

Am I right?

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Before re-requesting review, address unresolved/unanswered issues!

Plus consider improvements below.

File: iOS-School-02_stack.playground/Contents.swift [hunk 13]
```diff
@@ -1,20 +1,57 @@
+// MARK: - Stack
+
+/**
+ Provides a standard interface for getting, adding, and removing elements of a collection,
+ which is organized according to the principle "last in, first out", called a
+ [stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)).
+ The stack can hold `Int` values only.
+ You can add, remove an element from the stack and peek the element that added last.
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:8]: I would remove this line.

File: iOS-School-02_stack.playground/Contents.swift [hunk 14]
```diff
@@ -1,20 +1,57 @@
...
+ */
 class Stack {
+    /// Contains elements of the stack.
     private var stackElements = [Int]()
 
+    /**
+     Adds the element to the stack.
+     - Parameter element; An element to add on the stack.
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:16]:
```suggestion
     - Parameter element: An element to add on the stack.
```

File: iOS-School-02_stack.playground/Contents.swift [hunk 15]
```diff
@@ -1,20 +1,57 @@
...
+/**
+ Provides a standard interface for getting, adding, and removing elements of a collection,
+ which is organized according to the principle "last in, first out", called a
+ [stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)).
+ The stack can hold `Int` values only.
+ You can add, remove an element from the stack, peek the element that added last
+ and get the minimum value in the stack.
+ */
 class StackStatistics: Stack {
+    /// Collection of elements.
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:54]: Is that random collection of elements?

Could you please describe what `currentMinElements` stores in Russian?

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Consider improvements below.

Also please rename PR. E.g.: "HW#2: add documentation for Stack/StackStatistics classes".

File: iOS-School-02_stack.playground/Contents.swift [hunk 16]
```diff
@@ -1,20 +1,54 @@
...
+// MARK: - StackStatistics
+
+/**
+ Provides a standard interface for getting, adding, removing, and get the minimum
+ value element of a collection which is organized according to the principle "last in, first out", called a
+ [stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)).
+ The stack can hold `Int` values only.
+ */
 class StackStatistics: Stack {
+    /// Collection of current min element.
```
Comment (dersim-davaod) [inline iOS-School-02_stack.playground/Contents.swift:51]: Could you please translate this into Russian?
Will that be `коллекция текущего минимального элемента`?

Comment (dersim-davaod) [reply iOS-School-02_stack.playground/Contents.swift:51]: I would write something like:
```swift
/**
 Contains current min elements. Each time a new element is added into the stack, new current min element is added into the `currentMinElements ` collection. As soon as some element is removed from the stack, latest element is removed from the `currentMinElements` collection. Thus the latest element points to the min element in the stack at the current moment.
 */
```

Does that make sense?

**Review by dersim-davaod — CHANGES_REQUESTED** [review]

**Review by dersim-davaod — APPROVED** [review]
LGTM

**Pull request merged**